From 97f208aa50beb17075f5286668499c20afcb69cc Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Wed, 25 Sep 2013 16:31:12 +0700
Subject: Dalvik: Trace collector should be aware about big offset

BZ: 140700

Trace description data structure has u2 field to keep and offset of the bytecode
with respect to method start. However it is possible that offset is bigger than
0xFFFF. As a result during conversion offset is rounded and becomes incorrect.
After that JIT compiler is able to compile non-existing trace.

The patch stops the trace collection if offset becomes too big.

Category: device-enablement
Domain: AOSP-Dalvik-Interpreter
Origin: internal
Upstream-Candidate: yes

Change-Id: If32ae1b72c4c3ca6c5abfb3c7ecd7eb7b5797c7f
Orig-MCG-Change-Id: I636760e53e82bd1ab1d5cf39daef359ad96621f8
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Reviewed-on: http://android.intel.com:8080/136467
Reviewed-by: Beyler, Jean Christophe <jean.christophe.beyler@intel.com>
Reviewed-by: lab_aqa <lab_aqa@intel.com>
Reviewed-by: Provodin, Vitaly A <vitaly.a.provodin@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: Banerji, Udayan <udayan.banerji@intel.com>
Reviewed-by: Hartley, Timothy D <timothy.d.hartley@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 vm/interp/Jit.cpp |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/vm/interp/Jit.cpp b/vm/interp/Jit.cpp
index 473e490..16ec124 100644
--- a/vm/interp/Jit.cpp
+++ b/vm/interp/Jit.cpp
@@ -826,13 +826,23 @@ void dvmCheckJit(const u2* pc, Thread* self)
                 break;
             }
 
+            offset = lastPC - self->traceMethod->insns;
+            if (offset < 0 || offset > 0xFFFF)
+            {
+#if defined(SHOW_TRACE)
+                ALOGD("TraceGen: ending on %s, too big offset %x",
+                     dexGetOpcodeName(decInsn.opcode), offset);
+#endif
+                self->jitState = kJitTSelectEnd;
+                break;
+            }
+
 #if defined(SHOW_TRACE)
             ALOGD("TraceGen: adding %s. lpc:%#x, pc:%#x",
                  dexGetOpcodeName(decInsn.opcode), (int)lastPC, (int)pc);
 #endif
             flags = dexGetFlagsFromOpcode(decInsn.opcode);
             len = dexGetWidthFromInstruction(lastPC);
-            offset = lastPC - self->traceMethod->insns;
             assert((unsigned) offset <
                    dvmGetMethodInsnsSize(self->traceMethod));
             if (lastPC != self->currRunHead + self->currRunLen) {
-- 
1.7.4.1

